sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedSource.java

```diff
@@ -134,6 +136,106 @@ public Double getFractionConsumed() {
+     * <li>Any {@link BoundedReader reader} that has only returned its first element (aka,
+     * has never had a call to {@link #advance}) has a consumed parallelism of 0: the first element
+     * is the current element and is still being processed.
+     * <li>When processing record #30 (starting at 1) out of 50 in a perfectly splittable 50-record
+     * input, this value should be 29.
+     * <li>In a block-compressed value consisting of 5 blocks, the value should stay at 0 until the
+     * first record of the second block is returned; stay at 1 until the first record of the third
+     * block is returned, etc. Once the end-of-file is reached then the fifth block has been
+     * consumed and the value should stay at 5.
+     * <li>For an empty reader (in which the call to {@link #start} returned false), the
+     * consumed parallelism is 0.
+     * <li>For a non-empty, finished reader (in which the call to {@link #start} returned true and
+     * a call to {@link #advance} has returned false), the value returned must be at least 1
+     * and should equal the total parallelism in the source. For the two examples above, the
+     * 50-record file should have a consumed parallelism of 50 and the 5-block file should have
+     * a consumed parallelism of 5.
+     * </ul>
+     *
+     * <p>A reader that is implemented using a {@link RangeTracker} is encouraged to use the
+     * range tracker's ability to count split points to implement this method. See
+     * {@link OffsetBasedSource.OffsetBasedReader} and {@link OffsetRangeTracker} for an example.
+     *
+     * <p>Defaults to {@link #PARALLELISM_UNKNOWN}. Any value less than 0 will be interpreted
+     * as unknown.
+     *
+     * <h3>Thread safety</h3>
+     * See the javadoc on {@link BoundedReader} for information about thread safety.
+     *
+     * @see #getParallelismRemaining()
+     */
+    public long getParallelismConsumed() {
+      return PARALLELISM_UNKNOWN;
+    }
+
+    /**
+     * Returns the total amount of parallelism in the unprocessed part of this reader's current
+     * {@link BoundedSource} (as would be returned by {@link #getCurrentSource}). This corresponds
+     * to all unprocessed split point records (see {@link RangeTracker}), including the current
+     * record, in the remainder part of the source.
+     *
+     * <p>This function should be implemented only <strong>in addition to
+     * {@link #getParallelismConsumed()}</strong> and only if <em>an exact value can be
+     * returned</em>.
+     *
+     * <p>Consider the following two examples. An input that can be read in parallel down to the
+     * individual records, such as {@link CountingSource#upTo}, is called "perfectly splittable".
+     * Conversely, an example of a non-perfectly splittable input is a block-compressed file format
+     * such as {@link AvroIO}, in which a block of records has to be read as a whole, but different
+     * blocks can be read in parallel.
```

Comment:
Given how interleaved these are, it would be nice if this Javadoc could be in a single, nice place. What would you think of creating an interface with just these two methods, and putting the details there of how these methods interact. Then you can implement the interface and provide the defaults here?